src/cdomains/floatDomain.ml

```diff
@@ -0,0 +1,821 @@
+    | _, Bot -> false
+    | Interval (l1, h1), Interval (l2, h2) -> l1 >= l2 && h1 <= h2
+
+  let join v1 v2 =
+    match v1, v2 with
+    | Top, _ | _, Top -> Top
+    | Bot, v | v, Bot -> v
+    | Interval (l1, h1), Interval (l2, h2) -> Interval (min l1 l2, max h1 h2)
+
+  let meet v1 v2 =
+    match v1, v2 with
+    | Bot, _ | _, Bot -> Bot
+    | Top, v | v, Top -> v
+    | Interval (l1, h1), Interval (l2, h2) ->
+      let (l, h) = (max l1 l2, min h1 h2) in
+      if l <= h
+      then Interval (l, h)
+      else Bot
+
+  (** [widen x y] assumes [leq x y]. Solvers guarantee this by calling [widen old (join old new)]. *)
+  let widen v1 v2 = (**TODO: support 'threshold_widening' option *)
+    match v1, v2 with
+    | Top, _ | _, Top -> Top
+    | Bot, v | v, Bot -> v
+    | Interval (l1, h1), Interval (l2, h2) ->
+      (**If we widen and we know that neither interval contains +-inf or nan, it is ok to widen only to +-max_float,
+         because a widening with +-inf/nan will always result in the case above -> Top *)
+      let low = if l1 <= l2 then l1 else Float_t.lower_bound in
+      let high = if h1 >= h2 then h1 else Float_t.upper_bound in
+      norm @@ Interval (low, high)
+
+  let narrow v1 v2 =
+    match v1, v2 with (**we cannot distinguish between the lower bound beeing -inf or the upper bound beeing inf. Also there is nan *)
+    | Bot, _ | _, Bot -> Bot
+    | Top, _ -> v2
+    | Interval (l1, h1), Interval (l2, h2) ->
+      let low = if l1 = Float_t.lower_bound then l2 else l1 in
+      let high = if h1 = Float_t.upper_bound then h2 else h1 in
+      norm @@ Interval (low, high)
+    | Interval _, Top -> v1
+
+  (** evaluation of the unary and binary operations *)
+  let eval_unop onTop eval_operation op =
+    if is_top op then
+      Messages.warn ~category:Messages.Category.Float ~tags:[CWE 189; CWE 739]
+        "Operand of unary expression could be +/-infinity or Nan";
+    match op with
+    | Bot -> raise (ArithmeticOnFloatBot (Printf.sprintf "unop %s" (show op)))
+    | Interval v -> eval_operation v
+    | Top -> onTop
+
+  let eval_binop eval_operation op1 op2 =
+    if is_top op1 then
+      Messages.warn ~category:Messages.Category.Float ~tags:[CWE 189; CWE 739]
+        "First operand of arithmetic operation could be +/-infinity or Nan";
+    if is_top op2 then
+      Messages.warn ~category:Messages.Category.Float ~tags:[CWE 189; CWE 739]
+        "Second operand of arithmetic operation could be +/-infinity or Nan";
+    match (op1, op2) with
+    | Bot, _ | _, Bot -> raise (ArithmeticOnFloatBot (Printf.sprintf "%s op %s" (show op1) (show op2)))
+    | Interval v1, Interval v2 ->
+      let is_exact_before = is_exact op1 && is_exact op2 in
+      let result = norm @@ eval_operation v1 v2 in
+      (match result with
+       | Interval (r1, r2) when not (is_exact result) && is_exact_before ->
+         Messages.warn
+           ~category:Messages.Category.Float
+           ~tags:[CWE 197; CWE 681; CWE 1339]
```

Comment:
Are [CWE 197](https://cwe.mitre.org/data/definitions/197.html) and [CWE 681](https://cwe.mitre.org/data/definitions/681.html) really applicable for binary operations on floats?